src/deque.rs

```diff
@@ -0,0 +1,261 @@
+    offset: isize,
+}
+
+#[derive(Debug)]
+pub struct BackGuardInner<S> {
+    offset: isize,
+    pub inner: S,
+}
+
+// You'll notice that InnerQueue functions tend to take a MutexGuard as an
+// argument, rather than managing the mutex by themselves. This is done because
+// upstream in Sender we need to be sure that _multiple_ operations to Queue
+// happen isolated from other Senders, the Receiver on occasion. It's a little
+// tedious but since Rust mutex is tied to scope what else are you gonna do?
+impl<T, S> InnerQueue<T, S>
+where
+    S: ::std::default::Default,
+{
+    pub fn with_capacity(capacity: usize) -> InnerQueue<T, S> {
+        assert!(capacity > 0);
+        let mut data: Vec<*const T> = Vec::with_capacity(capacity);
+        for _ in 0..capacity {
+            data.push(ptr::null());
+        }
+        let raw_data = (&mut data).as_mut_ptr();
+        mem::forget(data);
+        InnerQueue {
+            capacity: capacity,
+            data: raw_data,
+            size: AtomicUsize::new(0),
+            back_lock: Mutex::new(BackGuardInner {
+                offset: 0,
+                inner: S::default(),
+            }),
+            front_lock: Mutex::new(FrontGuardInner { offset: 0 }),
+            not_empty: Condvar::new(),
+        }
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.capacity
+    }
+
+    pub fn size(&self) -> usize {
+        self.size.load(Ordering::Relaxed)
+    }
+
+    pub fn lock_back(&self) -> MutexGuard<BackGuardInner<S>> {
+        self.back_lock.lock().expect("back lock poisoned")
+    }
+
+    pub fn lock_front(&self) -> MutexGuard<FrontGuardInner> {
+        self.front_lock.lock().expect("front lock poisoned")
+    }
+
+    pub unsafe fn push_back(
+        &self,
+        elem: T,
+        guard: &mut MutexGuard<BackGuardInner<S>>,
+    ) -> Result<bool, Error<T>> {
+        let mut must_wake_dequeuers = false;
+        if !(*self.data.offset((*guard).offset)).is_null() {
+            return Err(Error::Full(elem));
+        } else {
+            *self.data.offset((*guard).offset) = Box::into_raw(Box::new(elem));
+            (*guard).offset += 1;
+            (*guard).offset %= self.capacity as isize;
+            if self.size.fetch_add(1, Ordering::Release) == 0 {
+                must_wake_dequeuers = true;
+            };
+        }
+        Ok(must_wake_dequeuers)
+    }
+
+    pub unsafe fn pop_back_no_block(&self, guard: &mut MutexGuard<BackGuardInner<S>>) -> Option<T> {
+        if self.size.load(Ordering::Acquire) == 0 {
+            None
+        } else {
+            if ((*guard).offset - 1) < 0 {
+                (*guard).offset = (self.capacity - 1) as isize;
+            } else {
+                (*guard).offset -= 1;
+            };
+            let elem: Box<T> = Box::from_raw(*self.data.offset((*guard).offset) as *mut T);
+            *self.data.offset((*guard).offset) = ptr::null_mut();
+            self.size.fetch_sub(1, Ordering::Release);
```

Comment:
This needs to be moved before any of the other pointer modifications, else we might end up allowing Receiver to pull something that is about to be popped off.